Add defstruct macro

rutenkolk opened the pull request:
~~[Note: I would consider this a draft PR, as I have not yet added tests (which could unearth some bugs and necessitate appropriate fixes).]~~

Hi, this PR contains the addition of a `defstruct` macro. It does the following:

* It adds serialize and deserialize code to a serde "registry" for the new type (details below)
* It generates a new type that has the specified members (details below)
* It adds an implmentation for `c-layout`
* It adds inline implementations for both `deserialize-from` and `serialize-into`
* It adds an implementation for `clojure.pprint/simple-dispatch`

## serde registry
The "registry" is implemented via the multimethods `generate-deserialize` and `generate-serialize` which produce code to de/serialize the respective types. This removes indirection in the de/serialize code for types that use other types. i think in the original discussion we were on the same page, but thought the other meant something different. The `defstruct` macro adds implementations to the multimethods for the newly generated type.

## the generated type
The new type is generated via `deftype` in the private function `generate-struct-record`. This is an attempt to strike a middle ground between the two positions of the original discussion, although the result might be a bit odd:

* The type implements both `IPersistentVector` **and** `IPersistentMap`. 
  * The basic idea is: if it is treated like a vector, it behaves like a vector. if it is treated like a map, it behaves like a map.
* It therefore implements both vector-like methods like `nth` as well as map-like methods like `without` (for e.g. `dissoc`). 
* If there is a an overlap in map/vector interface such as with `assoc`, it supports both paradigms of indices-as-keys and membernames-as-keys. Practically speaking, if you use something like `assoc` with a number as a key, it behaves like a vector (and will return a vector), otherwise like a map (and will return a map).
* one notable exception here is `foreach` which can't support both paradigms, and it is therefore implemented as if it's a vector. The rationale here is that the value of the type is composed of the actual values of the members, not the associated names of the places of the values. If you `map` or `reduce` over an object of this type, you will do so over the values of the members.
  
## with-c-layout
There was one implementation problem. Since padding was needed to be taken into account to allow for inline serdes, the new code for the macro needed to rely on `with-c-layout`. The problem here is that `with-c-layout` is in the `layout` namespace which already depends on `mem`. As a stopgap solution i simply copied the function over as a private function. I would be in favor of actually deprecating the `layout` namespace. for backwards compatibility the `with-c-layout` function in `layout` could depend on the one in `mem`. Not only is the `layout` namespace at this point somewhat anemic, it has also caused me trouble. I'm not sure if it's a bug, but due to `with-c-layout` being in `layout`, i ran into the problem that there are now two different `:padding` keywords, which i found confusing.

## tests & benchmarks
No tests or benchmarks exist right now. I don't *expect* the custom type to be slower than defrecord, but I want to test it.
Similarly, i do want to add a first set of tests for the de/serialization.

IGJoshua commented:
I absolutely love this. You've done a fantastic job! I look forward to seeing the tests that you add for this, and I'm also thinking ahead to reorganizing some of the existing serde code to use the new generate-serialize and generate-deserialize to add some inline arities to `serialize` and `deserialize` when the type argument is a constant. Don't worry about any of that in this PR, it's just something I want to use this for in the future.

IGJoshua commented:
So I like the way you've chosen to introduce the type registry. It integrates well with the existing tools, and provides a way to do inline arities for serialize and deserialize in the future. One hesitation I have at the moment looking over the code is the use of `::mem/array` to refer to actual java arrays.

Currently in coffi the `::mem/array` type serializes anything seqable, which *does* include arrays, and it deserializes to a vec.
I think that to support the direction you're going here, we should add optional kwargs as options in the array type, with a `:raw? true` option meaning that it will deserialize to a JVM array and will assume that the argument is an array, and then add some conditionals in to ensure that we have the fast path for array serialization.

## with-c-layout
For the `with-c-layout` problem, I think there's a couple things to be done. To start with, we can make the private version in coffi.mem use `:coffi.layout/padding` explicitly which doesn't require the namespace be loaded, which reduces it to just one padding key. Then for the rest, I'm a little undecided about it.

All the structs being passed over the C abi will most certainly use the `with-c-layout` layout, however the intention behind having the namespace in the first place was to allow easily serializing clojure maps into e.g. std140 or std430 from the GLSL spec, I just haven't gotten around to implementing those yet as I was wanting to get a `defstruct` macro and some codegen for an opengl bindgen library first.

Specifically though, if we remove the `coffi.layout` namespace and just assume everything is the c-layout, that will then mean there's no way for a user of the library to reach lower in the abstractions to implement a different layout for their usecase except to re-implement `defstruct` for their own layout.

rutenkolk commented:
I've been hacking away at this more.

I really like the idea of the optional keyword. I don't know what the default for the `raw` option should be. I've named it `raw-arrays?` for now, but i can switch it to the original `raw`. For now I defaulted to `raw-arrays? true`, which i can obviously change.
In any case, both versions work, also if mixed, like one defstruct using raw arrays and another one not using it but referring to the first defstruct.

Regarding the generated type: I've ran into issues with more implementations of the IPersistentMap interface colliding with IPersistentVector than i anticipated (like `keys` and `vals` for example relying on very specific `seq` behavior). I've defaulted to complete map like behavior if needed and removed the `IPersistentVector` interface from the type. But i've also still implemented the vector methods, but named differently, exposed via a new interface `coffi.mem.IStructImpl`. There is a new `StructVecWrap` type which implements `IPersistentVector` instead of `IPersistentMap` and which can take such a `IStructImpl` and offer the vector API with very little overhead, calling the vector methods of `IStructImpl`. to convert a type constructed with `defstruct` to this efficient vector implementation, and back, there are the new functions `as-vec` and `as-map`. The performance is really good, even with the indirection being as quick or quicker than native vectors in pretty much all cases. Similar optimisations have been done to generating seqs of the map or vector version of the type, being faster than seqs on maps and vectors respectively. I'll follow up with some benchmarks

the serde registry is unchanged for now, i'll have to think about nicely making them hygenic. it would definitely make things more complicated.

i also don't have a solution for the `with-c-layout` problem. one thing that confused me a bit from your answer though, is that ::layout/padding seems to be intentionally used in `coffi.layout`. Since there are also implementations in `coffi.mem` for `c-layout`, `serialize-into` and `deserialize-from` for `::padding` (`::mem/padding`) i was under the impression that `::layout/padding` was a typo and `::mem/padding` was the "correct" version of that keyword. since lots of code uses namespaced keywords i would strongly prefer deciding on one.

I still have to clean up a bit, and the above mentioned things aren't addressed yet, but functionality wise, i think things are mostly done and tested.

attached is a file with some random criterium runs that i've done to test the speed of stuff: results are promising. the new defstruct is faster than records, maps and sometimes vectors and seem to improve the speed of calling native functions with struct arguments / return types.

[mini-assorted-benchmarks.txt](https://github.com/user-attachments/files/17579320/mini-assorted-benchmarks.txt)

IGJoshua commented:
I've been following along as you've been adding commits, and it's been really fun to watch as it grows. Thanks for the wrap-up, it's helped clear everything up in my head though, since I don't usually haven enough time to fully dig into and understand the diffs when I do that.

I think I'd like the `:raw-arrays?` to default to false just to comply with the default behavior elsewhere, since a goal of coffi is for calling native code to feel like calling Clojure code. I think the naming is good if you put the argument on the struct itself, I had originally been thinking about a `[::mem/array ::whatever-type 3 :raw? true]` type extension to say that the native array will deserialize to a raw java array, similar to how the `[::mem/fn [] ::mem/void :raw-fn? true]` type argument disables automatic serdes on the argument and return types of the function as it crosses native boundaries. If we were to do it that way, we could collapse the coffi types for `vector-array` and `array` down into a single type, and it would even allow structs to have some arrays which are native and some which are not, which allows for more possible usecases. I can add support for these raw arrays to the basic types so you only have to worry about how they interact with structs.

I really like the solution of adding `as-vec` and `as-map` as functions you can call on the result to convert between the types, and I'm really happy to hear about the performance of this! I looked at the mini benchmarks and they looked really good.

If making the generation functions hygenic would be a significant amount of work then I would be comfortable with delaying that work and ensuring that the generation functions aren't a part of the public api of coffi for now.

So the `with-c-layout` the two different `padding` keywords are for different things. `::mem/padding` is a type which comes from the panama PaddingLayout which accounts for bytes without meaning just there to take up space. Then `::layout/padding` is a *field name* assigned to all padding "fields" within a struct. It's not intended to be kept or used for anything, and multiple padding fields get destructively written to the same key on structs as they're deserialized as it stands, and in the case of a `defstruct` implementation I'd actually want to just not have the padding field appear in the result at all, even though padding layout elements would still need to appear in the layout of the struct itself.

I'd like to thank you again for taking on this feature! I really like the work you're doing. :)

rutenkolk commented:
Thanks for the feedback! I'm happy things are shaping up somewhat nicely.

I think I simply didn't fully get how you meant the `:raw` option to work, but I really like the `[::mem/array ::whatever-type 3 :raw? true]` solution over `vector-array` (which i basically bolted on to make things work without touching the code for `::mem/array`, which i wasn't 100% happy with yet either).

I also think your rationale on `raw` or `raw-arrays?` to default to false is probably for the best. That does make sense in the context of the library. There are some practical concerns that i had myself once i tested things more extensively, most notably probably is that raw arrays don't play nicely with `=`.

Thank you for also clearing up that misunderstanding on the whole `:padding` situation! The reason i deal with the padding in the defstruct macro at all is to precompute where to (de)serialize the values are in the segment. I agree it shouldn't be emitted from the macro though.

also, i think i forgot to mention it, but i really like the idea of supporting different `layouts`, especially in the context of OpenGL / graphics development! so with all that information, the layout namespace makes a whole lot more sense to me!

i'll continue on the PR then and update when i hit what feels like the next milestone

rutenkolk commented:
Hi, I've opened this PR for review, since I tested and benchmarked everything.

The macro itself doesn't have a `raw?` option anymore, but it supports arrays with a `:raw? true` optional keyword like this: `[::mem/array ::whatever-type 3 :raw? true]`.

There were a few interesting things that happened along the way. One issue were the inline versions of the `write-x` functions like `write-int`. Sometimes reflection would occur but even `*warn-on-reflection* true` wouldn't catch it. Since writing this macro version with proper typehinting is finicky i added a new `with-typehints` macro to make it more bulletproof and applied it to all of the functions where it was causing issues. I can write this out by hand if needed, but it's very uncomfortable. In essence you either couldn't pass in a literal or you a form and have it be typehinted correctly as a primitive local before. Now this works, given that the local is actually of the right primitive type.

the internal `with-c-layout` function still exists, but it's using the keywords just like the original version in `coffi.layout`.

the macros are now hygenic, as in: they take in the expression that represents e.g. a segment and don't just use a random symbol name you have to match on the call-site.

i have spent some time optimizing serializing and deserializing and benchmarked everything. especially interesting is the story around arrays. some key takeaways: 

* if one has to decide between inlining and unrolling a loop, inlining is faster for raw arrays, but not for vectors
* with big enough arrays of primitives, it's advantageous for serializing to first create one big array and then copy that array with one call. this is apparently not true for deserializing, which is kind of surprising! A transient vector wins this case.
* there are tradeoffs when which method to read / write becomes the better option and the code auto-chooses the best, based on my benchmarks. this may need to be tested on multiple platforms though and adjusted accordingly.

![grafik](https://github.com/user-attachments/assets/90cc75de-2ee9-411c-9b64-be1e0ac24967)

![grafik](https://github.com/user-attachments/assets/16cbf1d7-769f-4aad-8433-aafbd774797e)

In any case, the "auto" version wich chooses the method performs as good or better than the best alternative for arrays and vectors respectively. That being said, raw arrays remain incredibly fast and small vectors (< 16 elements) are still pretty fast.

IGJoshua commented on src/clj/coffi/mem.clj:
```diff
@@ -1350,3 +1766,523 @@
+                      ::c-string 'String}]
+    (cond (and arr? raw?) (get array-types _type 'objects)
+          (and arr?)      `clojure.lang.IPersistentVector
+          :default        (get single-types _type (keyword (str *ns*) (str _type))))))
```
What does this default do? Calling `coffitype->typename` with something not in the specific lists here would result in something like `:some.ns/:some.ns/foo` which is unreadable.

Since this gets used to create tags for typehinting, I think that this only makes much sense if it's either the struct typename of another struct object, or one of these primitives. Should this do something to only produce a value in those cases?

rutenkolk replied:
oh, i believe this is a remnant from when the defstruct macro didn't use a namespaced keyword internally, so this was supposed to add the namespace to the typename. since the input here is always a correctly namespaced keyword anyway (if not some primitive), this can get simplified to `(get single-types type type)`. it didn't cause any problems so i didn't notice it. fixed in [cb3e620](https://github.com/IGJoshua/coffi/pull/14/commits/cb3e62057ffa1606f4e66ef3df8d6654deaa295e)

IGJoshua commented on src/clj/coffi/mem.clj:
```diff
@@ -1350,3 +1766,523 @@
+    :coffi.mem/float  `read-floats
+    :coffi.mem/double `read-doubles} _type))
+
+(defmulti  generate-deserialize (fn [& xs] (if (vector? (first xs)) (first (first xs)) (first xs))))
```
`ffirst` is a cool function :)

rutenkolk replied:
fair point. fixed in [10b8baa](https://github.com/IGJoshua/coffi/pull/14/commits/10b8baafa2f09909aaf98f967ca1a5b06f11b97c)

IGJoshua commented on src/clj/coffi/mem.clj:
```diff
@@ -1350,3 +1766,523 @@
+(defmethod generate-deserialize :coffi.mem/float    [_type offset segment-source-form] `(read-float   ~segment-source-form ~offset))
+(defmethod generate-deserialize :coffi.mem/double   [_type offset segment-source-form] `(read-double  ~segment-source-form ~offset))
+(defmethod generate-deserialize :coffi.mem/pointer  [_type offset segment-source-form] `(read-address ~segment-source-form ~offset))
+(defmethod generate-deserialize :coffi.mem/c-string [_type offset segment-source-form] (list `.getString (list `.reinterpret (list `.get (with-meta segment-source-form {:tag 'java.lang.foreign.MemorySegment}) `pointer-layout offset) `Integer/MAX_VALUE) 0))
```
Was there a specific thought that led to this version as opposed to the quasiquoted version? Nothing wrong with it, just curious about the thought process.
```suggestion
(defmethod generate-deserialize :coffi.mem/c-string [_type offset segment-source-form]
  `(.getString (.reinterpret (.get ~(with-meta segment-source-form {:tag 'java.lang.foreign.MemorySegment}) pointer-layout ~offset) Integer/MAX_VALUE) 0))
```

rutenkolk replied:
not really. i *think* this happened because at some point some metadata got swallowed up, and i suspected it ended up on a form or symbol in a macro and not on the emitted form and i wrote it explicitly while eliminating sources of error and never rewrote this. i agree on the syntax quoted version if only for uniformity.

IGJoshua commented on src/clj/coffi/mem.clj:
```diff
@@ -1350,3 +1766,523 @@
+(defn- generate-deserialize-array-as-vector-inline [array-type n offset segment-source-form]
+  (vec (map #(generate-deserialize array-type (+ offset (* (size-of array-type) %)) segment-source-form) (range n))))
+
+(defn- generate-deserialize-array-as-vector [array-type n offset segment-source-form]
+  (let [inline-cutoff 64] ;this magic value has been benchmarked, but it may need adjusting for specific architectures
+    (if (< n inline-cutoff)
+      (generate-deserialize-array-as-vector-inline array-type n offset segment-source-form)
+      (generate-deserialize-array-as-vector-loop array-type n offset segment-source-form))))
```
This doesn't use the `bulk` one. Was the conclusion that for vectors the bulk version is never best? If so is the reason to keep it in the code just that we might need to adjust these after further benchmarking?

rutenkolk replied:
initially, bulk copy was surprisingly good for vectors, but i profiled, noticed that vector creation with a backing array took surprisingly long, switched to transients, benchmarked it again, and now bulk copy is never the best option:

![grafik](https://github.com/user-attachments/assets/d7725d84-c5e5-4d97-ab73-39442d2de820)

as you can see, the inline version without a backing array outperforms other versions including bulk copy until about 64 elements, after which it has no benefit over the looping deserialize that doesn't use arrays (the latest version). In case of performance parity I would prefer the looping deserialize, as the method code for the generated `deserialize-from` can quickly grow in size, especially if we have types that depend on other types and thus include their deserialization code.

so yes the bold bulk copy for vectors is in there mostly for benchmarking purposes. as it is now, it seems strictly inferior to the other ways to deserialize native arrays to vectors.

rutenkolk replied:
update on this: no, for large enough arrays the gap only widens:

![grafik](https://github.com/user-attachments/assets/f38f390e-15a2-4c1a-81be-88d55fb13f6e)

rutenkolk replied:
removed the function in [d2afb2b](https://github.com/IGJoshua/coffi/pull/14/commits/d2afb2b1704b6e20826300e8bc0a6189ebe423df)

IGJoshua commented on src/clj/coffi/mem.clj:
```diff
@@ -1350,3 +1766,523 @@
+   (partition 2 2 (interleave (reductions + 0 (map (comp size-of second) fields)) fields))
+   (filter (fn [[_ [_ field-type]]] (not (and (vector? field-type) (= "padding" (name (first field-type)))))))))
+
+(defn register-new-struct-deserialization [typename [_struct fields]]
+  (let [typelist (typelist typename fields)]
+    (defmethod generate-deserialize typename [_type global-offset segment-source-form]
+      (->> typelist
+           (map-indexed
+            (fn [index [offset [_ field-type]]]
+              (generate-deserialize field-type (+ global-offset offset) segment-source-form)))
+           (cons (symbol (str (name typename) ".")))))))
```
I'd prefer if this returned the code for the defmethod to be included in the output of the `defstruct` macro, unless there's some specific reason why it should happen at macroexpansion time.

Same with the `register-new-struct-serialization` further down.

rutenkolk replied:
the rationale here was:

since we want to emit the inlined (de)serialization for the specific types, we need to register the code for that before we emit it. if there is a defstruct that contains another defstruct defined type as a member, `generate-deserialize` must have an implementation for that referenced type at macroexpansion time or else we cannot emit the code for the referenced type.

that is, unless i'm missing something here.

rutenkolk replied:
as discussed, for clarity, it will both happen at macroexpansion time, but also be emitted as per [1f5efb0](https://github.com/IGJoshua/coffi/pull/14/commits/1f5efb0982335bc93e4d27cbb5d39b8600aaf47e)

IGJoshua commented on src/clj/coffi/mem.clj:
```diff
@@ -1350,3 +1766,523 @@
+    (generate-deserialize-array-as-array array-type n offset segment-source-form)
+    (generate-deserialize-array-as-vector array-type n offset segment-source-form)))
+
+(defn- typelist [typename fields]
+  (->>
+   (partition 2 2 (interleave (reductions + 0 (map (comp size-of second) fields)) fields))
+   (filter (fn [[_ [_ field-type]]] (not (and (vector? field-type) (= "padding" (name (first field-type)))))))))
```
I'd prefer it if the field name came before the type rather than after for consistency with the `::mem/struct` type.

rutenkolk replied:
that is unfortunate. since i preferred it this way around, this isn't the only place where i made the assumption that the order is `typename member-name`, including the docstring of defstruct, defstruct itself (and by that extension the tests), typelist, and possibly other places. if this was a blocking change, it would probably need to be adjusted in all these places.

rutenkolk replied:
actually, i totally misread this. typelist is actually consistent with `::mem/struct`. `fields` is actually the exact vector as it appears in `::mem/struct`, since i need to check for the padding. as you can see in the inline function `fn [[_ [_ field-type]]]`, the first member of the tuple (the fieldname) is discarded and the second one (the coffi typename) is used.

what i refer to as `typename` here, is the name of the new struct type to be defined. funnily enough, something i totally missed as well, is that it's never used. i removed the argument and adjusted the calls to `typelist` in [06cd910](https://github.com/IGJoshua/coffi/pull/14/commits/06cd910279be75a1ce22a0cce13aaa856c4e050c)


other than that, the order of `typename fieldname` has been reversed to `fieldname typename` in [5a9d156](https://github.com/IGJoshua/coffi/pull/14/commits/5a9d156e1784b1f69f682038ffed7afca94f3e6f) and tests have been adjusted in [637f156](https://github.com/IGJoshua/coffi/pull/14/commits/637f1566639a5fb1c2414c3d188771b5d075def2)

IGJoshua commented on src/clj/coffi/mem.clj:
```diff
@@ -1350,3 +1766,523 @@
+(for-each-fixed-length 15)
+(for-each-fixed-length 16)
+
+(deftype struct-vec-iterator [^coffi.mem.IStructImpl struct-obj ^int size ^{:volatile-mutable true :tag int} i]
```
Just for consistency this should probably be upper camelcase, like other types.
```suggestion
(deftype StructVecIterator [^coffi.mem.IStructImpl struct-obj ^int size ^{:volatile-mutable true :tag int} i]
```
This would also require another line in `generate-struct-type` to change the constructor call.

rutenkolk replied:
I agree completely! fixed in [d24ebc7](https://github.com/IGJoshua/coffi/pull/14/commits/d24ebc7810a1c9b9b357173dee854384ac64657a)

IGJoshua commented on src/clj/coffi/mem.clj:
```diff
@@ -1350,3 +1766,523 @@
+            (struct-methods [] [(s-count) (s-containsKey) (s-valAt) (s-valAt-2) (s-entryAt)])
+            (prefix-methods [prefix ms] (map (fn [[method-name & tail]] (cons (with-meta (symbol (str prefix method-name)) (meta method-name)) tail)) ms))
+            (impl-methods [] (concat (prefix-methods "map_" (map-methods)) (prefix-methods "vec_" (vec-methods)) (prefix-methods "struct_" (struct-methods))))]
+      (concat
+       [`deftype (symbol (name typename)) (vec typed-member-symbols) `coffi.mem.IStruct `coffi.mem.IStructImpl `clojure.lang.IPersistentMap `clojure.lang.MapEquivalence `java.util.Map `clojure.lang.IFn]
+       (struct-methods)
+       (map-methods)
+       (impl-methods)
+       [(s-nth-key)
+        (invoke1)
+        (invoke2)
+        (list 'asMap ['this] 'this)
+        (list 'asVec ['this] (list `VecWrap. 'this))]))))
```
Not required, but it seems a bit more clear to me when using the normal quasiquote syntax.
```suggestion
      `(deftype ~(symbol (name typename)) ~(vec typed-member-symbols)
         coffi.mem.IStruct
         ~@(struct-methods)
         coffi.mem.IStructImpl
         ~@(impl-methods)
         clojure.lang.IPersistentMap
         clojure.lang.MapEquivalence
         java.util.Map
         ~@(map-methods)
         clojure.lang.IFn
         ~(s-nth-key)
         ~(invoke1)
         ~(invoke2)

         (~'asMap [~'this] ~'this)
         (~'asVec [~'this] (VecWrap. ~'this)))
```

rutenkolk replied:
I have no strong preference. I think your change looks very readable! I like the grouped Interfaces and then emitted method definitions!

IGJoshua reviewed (changes requested):
I've got a few specific things in the review I'd like resolved or to discuss, and attached here I've got a few patches that I'd like if they were applied to the PR.

[0004-Don-t-use-underscore-on-used-args.patch.txt](https://github.com/user-attachments/files/18287250/0004-Don-t-use-underscore-on-used-args.patch.txt)
[0003-Remove-duplicate-c-layout-implementation.patch.txt](https://github.com/user-attachments/files/18287251/0003-Remove-duplicate-c-layout-implementation.patch.txt)
[0002-Fix-warning-about-defstruct-redefinition.patch.txt](https://github.com/user-attachments/files/18287252/0002-Fix-warning-about-defstruct-redefinition.patch.txt)
[0001-Use-a-once-only-impl-rather-than-with-typehints.patch.txt](https://github.com/user-attachments/files/18287253/0001-Use-a-once-only-impl-rather-than-with-typehints.patch.txt)

rutenkolk commented:
I think work on this PR is nearing completion. As for the performance, here are some benchmark results:


### Serializing a struct with `n` amount of `::mem/int` members:

linear scale:
![grafik](https://github.com/user-attachments/assets/4f9989d9-bf29-4b39-a23a-dbccf79e3525)

logarithmic scale:
![grafik](https://github.com/user-attachments/assets/3c94cbb4-ba51-4425-80aa-1cd1327d685f)


### Deserializing a struct with `n` amount of `::mem/int` members:

linear scale:
![grafik](https://github.com/user-attachments/assets/5fba0622-4bca-4f6d-8e13-ccbb78fda016)

logarithmic scale:
![grafik](https://github.com/user-attachments/assets/12425cb4-d45a-4719-8741-dc12e6b8025b)


### Serializing a struct with one `::mem/array` of `::mem/int`s of fixed size `n`

all individual ways to serialize, logarithmic scale:
![grafik](https://github.com/user-attachments/assets/136aa14d-aed8-4859-99be-cdac5db566c1)

comparison defstruct with raw arrays and vectors for arrays vs. defalias, logarithmic scale:
![grafik](https://github.com/user-attachments/assets/e85d7481-908f-44f1-a982-1e6f031f49f6)

### Deserializing a struct with one `::mem/array` of `::mem/int`s of fixed size `n`

all individual ways to deserialize, logarithmic scale:
![grafik](https://github.com/user-attachments/assets/3753c849-2771-4c51-ad14-0b282d424e44)

comparison defstruct with raw arrays and vectors for arrays vs. defalias, logarithmic scale:
![grafik](https://github.com/user-attachments/assets/a5668088-ae38-432e-8381-80f7ac06aab6)


In all cases, a significant speedup has been achieved, often more than an order of magnitude, in special cases more than two. For native arrays, raw java arrays outperform vectors quite heavily. there may be room for improving this specific codepath, but it is still a noticeable improvement. For some cases, like raw arrays, the performance is in the realm of only a few nanoseconds and one of the biggest factors actually becomes the initial dispatch of the multimethod, so the actual time the de/serialization takes is probably hard to improve significantly further without changing aspects of how coffi itself operates.

rutenkolk commented:
to document a last touch:

i moved `with-c-layout` to `coffi.layout` again and `load-file`ed it in `coffi.mem` right above the definition of `defstruct` (the latest possible time, so that it hopefully will cause as little problems as possible, should `coffi.layout` be developed further).

defstruct can still be called from `coffi.mem` and even after removing the dependency on `coffi.layout` in `mem_test.clj` no test fails, so i think this worked out just fine!

rutenkolk commented:
i actually just pushed another fix here: the introduction of `load-file` in mem.clj caused the library to not find layout.clj when used as a dependency in other projects. the correct function to use is `load`. this was addressed in #[f307d1e](https://github.com/IGJoshua/coffi/pull/14/commits/f307d1ed5131487ad233dcbd509a9f71843bcd51)

rutenkolk commented:
🥳
